Update argument_value_resolver.rst

motivatedsloth opened the pull request:
I'm proposing 3 changes.
1. I added a note with a link to the @ParamConverter page. When the argument is a doctrine entity(as the current version of the page shows) the ArgumentValueResolver will not work. The ParamConverter must be used.
2. I updated the controller example to show the argument as the UserInterface instead of the User entity. This will work.
3. Added two missing use statements in the ArgumentValueResolver example.

motivatedsloth committed:
Update argument_value_resolver.rst

I'm proposing 3 changes.
1. I added a note with a link to the @ParamConverter page. When the argument is a doctrine entity(as the current version of the page shows) the ArgumentValueResolver will not work. The ParamConverter must be used.
2. I updated the controller example to show the argument as the UserInterface instead of the User entity. This will work.
3. Added two missing use statements in the ArgumentValueResolver example.

HeahDude commented on controller/argument_value_resolver.rst:
```diff
@@ -51,12 +55,12 @@ controller::
     class UserController
     {
-        public function indexAction(User $user)
+        public function indexAction(UserInterface $user)
```
Why this change? It's inconsistent with the example below.

motivatedsloth replied:
Good point. I'll update the example to make it consistent. Both the controller argument and the ArgumentValueResolver should use the UserInterface.

HeahDude replied:
I disagree, the implementation you suggest is already done in the security bundle as of 3.2 (see https://github.com/symfony/symfony/blob/3.2/src/Symfony/Bundle/SecurityBundle/SecurityUserValueResolver.php), so it would not be custom anymore.

motivatedsloth replied:
That is correct however the current example will not work. 
It throws the following exception "Unable to guess how to get a Doctrine instance from the request information for parameter "user"."
It's because as you noted the param converters execute first. The DoctrineParamConverter=>supports() returns true, because it is a doctrine entity. Then the DoctrineParamConverter=>apply() method then throws the exception.
The security bundle SecurityUserValueResolver is never executed.

HeahDude replied:
You are right, so the note should still explicit two things: the fact that converters are executed first and that using another placeholder than "id" in the route will avoid the automatic mapping from the framework extra bundle configuration.

motivatedsloth replied:
I'm trying to understand.
ParamConverters are executed first. 
ArgumentValueResolvers are executed later.
They both allow you to type-hint arguments in your controller.
 When should you use a ParamConverter and when would you use an ArgumentValueResolver?

motivatedsloth replied:
I think I figured it out. There are two issues. One is that it would be helpful to link to the ParamConverter page and clarify when that is appropriate. I'll update my pull request.
The other is a bug in the DoctrineParamConverter that is causing the current example to fail. I will report that bug too.

HeahDude commented on controller/argument_value_resolver.rst:
```diff
@@ -40,6 +40,10 @@ Symfony ships with four value resolvers in the HttpKernel component:
+    
+.. note::
+    If your argument is a Doctrine Entity you will need to create a 
+    :doc:`param converter </bundles/SensioFrameworkExtraBundle/annotations/converters>`
```
I think the note should explicit that param converters are executed before argument value resolvers.

This can be useful though since the entity is accessible through request attributes for any custom resolvers.

HeahDude reviewed (changes requested):
Thanks for that PR!

motivatedsloth committed:
Note about doctrine entities

Link to param converter page. 
Updated Resolver with UserInterface to be consistent with Controller
